### ISSUE TITLE
Jackd doesn't work with vc4hdmi

### ISSUE THREAD

**Issue by vanfanel** [raspberrypi/linux#4956]
### Describe the bug

Jackd2 doesn't work with vc4hdmi audio device.
Using default bullseye **vc4hdmi.conf**
I tried building latest jack2 and it fails too.

This is what happens:
```

~/src# jackd -d alsa --device default:vc4hdmi0
jackdmp 1.9.20
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2021 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
creating alsa driver ... default:vc4hdmi0|default:vc4hdmi0|1024|2|48000|0|0|nomon|swmeter|-|32bit
ALSA lib conf.c:5200:(snd_config_expand) Unknown parameters vc4hdmi0
ALSA lib control.c:1471:(snd_ctl_open_noupdate) Invalid CTL default:vc4hdmi0
control open "default:vc4hdmi0" (Invalid argument)
ALSA: Cannot open PCM device alsa_pcm for capture. Falling back to playback-only mode
configuring for 48000Hz, period = 1024 frames (21.3 ms), buffer = 2 periods
ALSA: final selected sample format for playback: 32bit float little-endian
You appear to be using the ALSA software "plug" layer, probably
a result of using the "default" ALSA device. This is less
efficient than it could be. Consider using a hardware device
instead rather than using the plug layer. Usually the name of the
hardwa
ALSA: use 2 periods for playback
Cannot bind server to socket err = No such file or directory
Server channel open error
JackServer::Open failed with -1
Failed to open server
```

Any idea on what's wrong, please?

### Steps to reproduce the behaviour

Try to launch jackd with the **vc4hdmi** audio device.

### Device (s)

Raspberry Pi 4 Mod. B, Raspberry Pi 400

### System

Raspberry Pi reference 2021-10-30
Generated using pi-gen, https://github.com/RPi-Distro/pi-gen, c12b1df4ed6416fb0df33ba1731c5b13c1bdbdf8, stage2

Linux 5.10.103-v8+ #1530 SMP PREEMPT Tue Mar 8 13:06:35 GMT 2022 aarch64 GNU/Linux

**Comment by HiassofT** [comment]
ALSA side is fine (except you should better use hdmi:vc4hdmi0 instead of the default PCM), jack fails because it can't bind to it's socket.

A quick test with jackd2 here looks fine
```
root@raspberrypi:~# jackd -d alsa -d hdmi:vc4hdmi0 -P
jackdmp 1.9.17
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2021 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
ALSA lib control.c:1471:(snd_ctl_open_noupdate) Invalid CTL hdmi:vc4hdmi0
control open "hdmi:vc4hdmi0" (No such file or directory)
ALSA lib control.c:1471:(snd_ctl_open_noupdate) Invalid CTL hdmi:vc4hdmi0
control open "hdmi:vc4hdmi0" (No such file or directory)
creating alsa driver ... hdmi:vc4hdmi0|-|1024|2|48000|0|0|nomon|swmeter|-|32bit
ALSA lib control.c:1471:(snd_ctl_open_noupdate) Invalid CTL hdmi:vc4hdmi0
control open "hdmi:vc4hdmi0" (No such file or directory)
configuring for 48000Hz, period = 1024 frames (21.3 ms), buffer = 2 periods
ALSA: final selected sample format for playback: 32bit integer little-endian
ALSA: use 2 periods for playback
```

I'd suggest you check into your jackd setup and/or ask on the forum, the issue you have is not related to the linux kernel.

**Comment by vanfanel** [comment]
@HiassofT Sorry if this is not the place to ask then. And thanks for taking the time to answer even if it's not the right place! Will ask this on the Raspberry Pi forum.

**Comment by vanfanel** [comment]
@HiassofT Maybe you can help me after all.
Running as the normal pi user, just like you did, I get this:
```

pi@minimal:/root $ jackd -d alsa -d hdmi:vc4hdmi0 -P
jackdmp 1.9.20
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2021 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
Cannot lock down 107350048 byte memory area (Cannot allocate memory)
creating alsa driver ... hdmi:vc4hdmi0|-|1024|2|48000|0|0|nomon|swmeter|-|32bit
ALSA lib control.c:1471:(snd_ctl_open_noupdate) Invalid CTL hdmi:vc4hdmi0
control open "hdmi:vc4hdmi0" (No such file or directory)
ALSA lib conf.c:5200:(snd_config_expand) Unknown parameters vc4hdmi0
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM hdmi:vc4hdmi0
ALSA: Cannot open PCM device alsa_pcm for playback. Falling back to capture-only mode
Cannot initialize driver
JackServer::Open failed with -1
Failed to open server
```

Can you please tell me why is it saying "ALSA: Cannot open PCM device alsa_pcm for playback"? 
I mean, what should be creating the "alsa_pcm" device?

**Comment by vanfanel** [comment]
@HiassofT I am reopening this because it seems I don't have ah hdmi:vc4hdmi0 device/card combination on my system.

Hence, a couple of questions arise:

Are you using ALSA emulation on top of PulseAudio, or direct ALSA?
Are you using the default vc4hdmi.conf in Bullseye?

**Comment by HiassofT** [comment]
I'm running plain RPiOS Bullseye lite on RPi4 here, no pulseaudio installed and jackd from jackd2 distro package runs fine both as root and pi user.

Have you checked if you can use the alsa device at all, eg with "speaker-test -c 2 -t sine -D hdmi:vc4hdmi0"? Do you see the hdmi pcm in "aplay -L" output?

**Comment by vanfanel** [comment]
@HiassofT **speaker-test** sine test on **hdmi:vc4hdmi0** succeeds here, too:

~# speaker-test -c 2 -t sine -D hdmi:vc4hdmi0

speaker-test 1.2.4
```

Playback device is hdmi:vc4hdmi0
Stream parameters are 48000Hz, S16_LE, 2 channels
Sine wave rate is 440,0000Hz
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 64 to 65536
Period size range from 32 to 32768
Using max buffer size 65536
Periods = 4
was set period_size = 16384
was set buffer_size = 65536
 0 - Front Left
 1 - Front Right
Time per period = 4,111098
 0 - Front Left
```
Both the sine and default noise tests on **hdmi:vc4hdmi0**  succeed only if I pass the channel number (-c 2).
Without the channel number, they both fail:

```
~# speaker-test -D hdmi:vc4hdmi0

speaker-test 1.2.4

Playback device is hdmi:vc4hdmi0
Stream parameters are 48000Hz, S16_LE, 1 channels
Using 16 octaves of pink noise
Channels count (1) not available for playbacks: Invalid argument
Setting of hwparams failed: Invalid argument
```

So, the natural next step for me was to tell jackd to use 2 channels with the ALSA driver, too:
```
~# jackd -d alsa -d hdmi:vc4hdmi0 -P -o 2
jackdmp 1.9.20
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2021 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
creating alsa driver ... hdmi:vc4hdmi0|-|1024|2|48000|0|2|nomon|swmeter|-|32bit
ALSA lib control.c:1471:(snd_ctl_open_noupdate) Invalid CTL hdmi:vc4hdmi0
control open "hdmi:vc4hdmi0" (No such file or directory)
configuring for 48000Hz, period = 1024 frames (21.3 ms), buffer = 2 periods
ALSA: final selected sample format for playback: 32bit integer little-endian
ALSA: use 2 periods for playback
Cannot bind server to socket err = No such file or directory
Server channel open error
JackServer::Open failed with -1
Failed to open server

```

Sadly, jackd fails with the same error even if I specify 2 channels.

As for **aplay -L**, this is what I see (I believe it's fine):
```

~# aplay -L
null
    Discard all samples (playback) or generate zero samples (capture)
default
    Default Audio Device
sysdefault
    Default Audio Device
hw:CARD=vc4hdmi0,DEV=0
    vc4-hdmi-0, MAI PCM i2s-hifi-0
    Direct hardware device without any conversions
plughw:CARD=vc4hdmi0,DEV=0
    vc4-hdmi-0, MAI PCM i2s-hifi-0
    Hardware device with all software conversions
default:CARD=vc4hdmi0
    vc4-hdmi-0, MAI PCM i2s-hifi-0
    Default Audio Device
sysdefault:CARD=vc4hdmi0
    vc4-hdmi-0, MAI PCM i2s-hifi-0
    Default Audio Device
hdmi:CARD=vc4hdmi0,DEV=0
    vc4-hdmi-0, MAI PCM i2s-hifi-0
    HDMI Audio Output
dmix:CARD=vc4hdmi0,DEV=0
    vc4-hdmi-0, MAI PCM i2s-hifi-0
    Direct sample mixing device
hw:CARD=vc4hdmi1,DEV=0
    vc4-hdmi-1, MAI PCM i2s-hifi-0
    Direct hardware device without any conversions
plughw:CARD=vc4hdmi1,DEV=0
    vc4-hdmi-1, MAI PCM i2s-hifi-0
    Hardware device with all software conversions
default:CARD=vc4hdmi1
    vc4-hdmi-1, MAI PCM i2s-hifi-0
    Default Audio Device
sysdefault:CARD=vc4hdmi1
    vc4-hdmi-1, MAI PCM i2s-hifi-0
    Default Audio Device
hdmi:CARD=vc4hdmi1,DEV=0
    vc4-hdmi-1, MAI PCM i2s-hifi-0
    HDMI Audio Output
dmix:CARD=vc4hdmi1,DEV=0
    vc4-hdmi-1, MAI PCM i2s-hifi-0
    Direct sample mixing device
```

Again, any ideas are VERY welcome. No hurries, it's for a long-term project, but your help is maybe the only hope I have with this.

**Comment by HiassofT** [comment]
Everything's still fine on the ALSA side, the error is still the same as in your first post, your self-compiled jackd can't bind to it's socket - you have to look into that.

I recommend you start with a fresh RPiOS lite installation, install jackd2 package from repository and then work down from there.

**Comment by vanfanel** [comment]
@HiassofT Thanks.
Looking at the RPiOS jackd2 package, it seems to depend on libDBUS.
Do you know if jackd2 uses DBUS for the socket stuff?
In other words, do you know if there's some external requeriment for the jackd sockets to work? (DBUS service running, etc)

**Comment by HiassofT** [comment]
Sorry, I'm not familiar with jack internals. I had a look at it once some 15 years ago and now only installed the jackd2 apt package but that's it.

As this is getting highly off topic and is no linux kernel issue I suggest again you better close this issue and seek for support on the RPi forums

**Comment by vanfanel** [comment]
@HiassofT Understood, sir.
Again, thanks a lot for your time on this, really.

**Comment by vanfanel** [comment]
@HiassofT and future me in case of need:

Git this to work. I was missing the /dev/shm directory mounted as tmps.
So I added this to my custom init script:
```

mkdir /dev/shm
mount -t tmpfs tmpfs /dev/shm
```

..and jackd2 started working.
Found this out by debugging the `JackServerSocket::Bind` method. Should have started there, instead of annoying the kernel people with this. Sorry and thanks once more.
